File: 10) Perceus - 2019/src/Core/Graphics/Entities/Camera.cpp
#include "Perceus/Core/Graphics/Entities/Camera.h"

#include "Perceus/Core/Engine.h"
#include "Perceus/Data/Inc.h"

namespace pcs
{
    Camera::Camera(Window* w, float FOV, float zNear, float zFar) :
            window(w), projection(), _FOV(FOV), _zNear(zNear), _zFar(zFar)
    {
        PS_CORE_DEBUG("Constructing Camera.");
        window->pushCamera(this);
        Camera::makeProjection(FOV, zNear, zFar);
        PS_CORE_DEBUG("Done with camera.");
    }
    
    void Camera::move(const Direction direction, const float speed)
    {
        switch(direction)
        {
            case Direction::Forward:
            {
                // cos rot.y changes x
                // sin rot.y changes z

                getLocation().x += std::cosf(getRotation().y + (90.f * 3.14159f / 180.f)) * speed;
                getLocation().z += std::sinf(getRotation().y + (90.f * 3.14159f / 180.f)) * speed;
                return;
            }

            case Direction::Backward:
            {
                // Same as forwards, but backwards
                getLocation().x -= std::cosf(getRotation().y + (90.f * 3.14159f / 180.f)) * speed;
                getLocation().z -= std::sinf(getRotation().y + (90.f * 3.14159f / 180.f)) * speed;
                return;
            }

            case Direction::Left:
            {
                getLocation().x -= std::cosf(getRotation().y) * speed;
                getLocation().z -= std::sinf(getRotation().y) * speed;
                return;
            }

            case Direction::Right:
            {
                getLocation().x += std::cosf(getRotation().y) * speed;
                getLocation().z += std::sinf(getRotation().y) * speed;
                return;
            }
        }
    }

    Mat4f Camera::getView() 
    {
        Mat4f t = Mat4f::makeTranslation(getLocation() * -1);
        Mat4f r = Mat4f::makeRotation(getRotation());
        return r * t;
    }

    void Camera::associateWindow(Window* win)
    {
        window = win;
    }

    void Camera::makeProjection(float FOV, float zNear, float zFar)
    {
        if (FOV == -1)   FOV = _FOV;
        if (zNear == -1) zNear = _zNear;
        if (zFar == -1)  zFar = _zFar;

        PS_CORE_DEBUG("Making proj");
        std::cout << window << "\n";
        Vec2u size = window->getSize();
        PS_CORE_DEBUG("Camera: Window Size {0}x{1}", size.x, size.y);
        float aspectRatio = (float)size.y / (float)size.x;

        // WTF is going on here, 10 degrees behaves as if it were 100 degrees
        float fovRad = FOV / 10.f * 3.14159f / 180.f;
        float invTan = 1.f / tanf(fovRad / 2.f);
        
        projection(0, 0) = aspectRatio * invTan;
        projection(1, 1) = invTan;
        projection(2, 2) = zFar / (zFar - zNear);
        projection(2, 3) = -1;
        projection(3, 2) = (-zFar * zNear) / (zFar - zNear);
    }
}

File: bin/camera.py
import math

class Camera:
    x = 0; y = 0; z = 0
    xrot = 0; yrot = 0;
    speed = .0005; distance = 75

    def get_xrot(self):
        return self.xrot / 250.0

    def sprint_on(self):
        #self.speed = .001
        self.speed  = 1

    def sprint_off(self):
        #self.speed = .0005
        self.speed = .1

    def get_yrot(self):
        return self.yrot / 250.0

    def get_loc(self):
        return (self.x, self.y, self.z)

    def forward(self):
        loc = list(self.get_loc())
        loc[0] = self.speed * -math.sin(self.get_xrot()) + loc[0]
        loc[2] = self.speed * -math.cos(self.get_xrot()) + loc[2]

        self.x = loc[0]; self.y = loc[1]; self.z = loc[2]

    def backward(self):
        loc = list(self.get_loc())
        loc[0] = self.speed * math.sin(self.get_xrot()) + loc[0]
        loc[2] = self.speed * math.cos(self.get_xrot()) + loc[2]

        self.x = loc[0]; self.y = loc[1]; self.z = loc[2]

    def left(self):
        loc = list(self.get_loc())
        loc[0] = self.speed * -math.sin(self.get_xrot() - (math.pi / 2.0)) + loc[0]
        loc[2] = self.speed * -math.cos(self.get_xrot() - (math.pi / 2.0)) + loc[2]

        self.x = loc[0]; self.y = loc[1]; self.z = loc[2]

    def right(self):
        loc = list(self.get_loc())
        loc[0] = self.speed * -math.sin(self.get_xrot() + (math.pi / 2.0)) + loc[0]
        loc[2] = self.speed * -math.cos(self.get_xrot() + (math.pi / 2.0)) + loc[2]

        self.x = loc[0]; self.y = loc[1]; self.z = loc[2]
